### PR TITLE
[clang] Add tests for Core issues about friend templates

### PR DESCRIPTION
**Pull request by Endilll**
This patch covers the following Core issues:
[CWG1862](https://cplusplus.github.io/CWG/issues/1862.html) "Determining “corresponding members” for friendship"
[CWG1918](https://cplusplus.github.io/CWG/issues/1918.html) "`friend` templates with dependent scopes"
[CWG1945](https://cplusplus.github.io/CWG/issues/1945.html) "Friend declarations naming members of class templates in non-templates"

All of them were resolved by CWG1862, where the current wording of [[temp.friend] p5](https://eel.is/c++draft/temp.friend#5) comes from:
> A template friend declaration may declare a member of a dependent type to be a friend[.](https://eel.is/c++draft/temp.friend#5.sentence-1) The friend declaration shall declare a function or specify a type with an [elaborated-type-specifier](https://eel.is/c++draft/dcl.type.elab#nt:elaborated-type-specifier), in either case with a [nested-name-specifier](https://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier) ending with a [simple-template-id](https://eel.is/c++draft/temp.names#nt:simple-template-id), C, whose [template-name](https://eel.is/c++draft/temp.names#nt:template-name) names a class template[.](https://eel.is/c++draft/temp.friend#5.sentence-2) The template parameters of the template friend declaration shall be deducible from C ([[temp.deduct.type]](https://eel.is/c++draft/temp.deduct.type))[.](https://eel.is/c++draft/temp.friend#5.sentence-3) In this case, a member of a specialization S of the class template is a friend of the class granting friendship if deduction of the template parameters of C from S succeeds, and substituting the deduced template arguments into the friend declaration produces a declaration that corresponds to the member of the specialization[.](https://eel.is/c++draft/temp.friend#5.sentence-4)

A useful example is provided after the paragraph: https://eel.is/c++draft/temp.friend#example-4.

Neither of Core issues is implemented, because we don't support dependent nested friend specifiers.

### CODE REVIEW

File: clang/test/CXX/drs/cwg18xx.cpp [hunk 1]
```diff
@@ -373,6 +373,98 @@ namespace cwg1837 { // cwg1837: 3.3
...
+  template<class T>
+  template<T U>
+  friend T A<T>::i();
+  // expected-warning@-1 {{dependent nested name specifier 'A<T>::' for friend class declaration is not supported; turning off access control for 'C'}}
+};
+
+C c;
+
+template<class T>
+void A<T>::B::e() { (void)c.private_int; }
```
Comment (shafik) [inline clang/test/CXX/drs/cwg18xx.cpp:445]: I guess these work b/c as the diagnostic above said "we turn off access control"?

File: clang/test/CXX/drs/cwg19xx.cpp [hunk 2]
```diff
@@ -121,6 +145,21 @@ derived d2(42, 9);
...
+  };
+};
+class X {
+  static int x;
+  // FIXME: this is ill-formed, because A<T>::B::C does not end with a simple-template-id
+  template <typename T>
+  friend class A<T>::B::C;
+  // expected-warning@-1 {{dependent nested name specifier 'A<T>::B::' for friend class declaration is not supported; turning off access control for 'X'}}
+};
+} // namespace cwg1918
```
Comment (shafik) [inline clang/test/CXX/drs/cwg19xx.cpp:161]:
```suggestion
} // namespace cwg1945
```

**Review by shafik — COMMENTED** [review]
LGTM after fixing minor comment

**Review by shafik — APPROVED** [review]

File: clang/test/CXX/drs/cwg18xx.cpp (hunk 1 above)
Comment (Endilll) [reply clang/test/CXX/drs/cwg18xx.cpp:445]: No, `A<T>::B` is a friend. `D::g` is the only function that doesn't have access in this example. Consult https://eel.is/c++draft/temp.friend#example-4 for details.

**Review by erichkeane — APPROVED** [review]

**Pull request merged**